presto-pinot/src/main/java/io/prestosql/pinot/PinotClusterInfoFetcher.java

```diff
@@ -0,0 +1,418 @@
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static io.airlift.http.client.StringResponseHandler.createStringResponseHandler;
+import static io.prestosql.pinot.PinotErrorCode.PINOT_INVALID_CONFIGURATION;
+import static io.prestosql.pinot.PinotErrorCode.PINOT_UNABLE_TO_FIND_BROKER;
+import static io.prestosql.pinot.PinotErrorCode.PINOT_UNEXPECTED_RESPONSE;
+import static io.prestosql.pinot.PinotMetrics.isValidPinotHttpResponseCode;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static org.apache.pinot.common.config.TableNameBuilder.extractRawTableName;
+
+public class PinotClusterInfoFetcher
+{
+    private static final Logger log = Logger.get(PinotClusterInfoFetcher.class);
+    private static final String APPLICATION_JSON = "application/json";
+    private static final Pattern BROKER_PATTERN = Pattern.compile("Broker_(.*)_(\\d+)");
+
+    private static final String GET_ALL_TABLES_API_TEMPLATE = "tables";
+    private static final String TABLE_INSTANCES_API_TEMPLATE = "tables/%s/instances";
+    private static final String TABLE_SCHEMA_API_TEMPLATE = "tables/%s/schema";
+    private static final String ROUTING_TABLE_API_TEMPLATE = "debug/routingTable/%s";
+    private static final String TIME_BOUNDARY_API_TEMPLATE = "debug/timeBoundary/%s";
+
+    private final PinotConfig pinotConfig;
+    private final PinotMetrics pinotMetrics;
+    private final HttpClient httpClient;
+
+    private final Ticker ticker = Ticker.systemTicker();
+
+    private final LoadingCache<String, List<String>> brokersForTableCache;
+
+    private final JsonCodec<GetTables> tablesJsonCodec;
+    private final JsonCodec<BrokersForTable> brokersForTableJsonCodec;
+    private final JsonCodec<RoutingTables> routingTablesJsonCodec;
+    private final JsonCodec<TimeBoundary> timeBoundaryJsonCodec;
+
+    @Inject
+    public PinotClusterInfoFetcher(
+            PinotConfig pinotConfig,
+            PinotMetrics pinotMetrics,
+            @ForPinot HttpClient httpClient,
+            JsonCodec<GetTables> tablesJsonCodec,
+            JsonCodec<BrokersForTable> brokersForTableJsonCodec,
+            JsonCodec<RoutingTables> routingTablesJsonCodec,
+            JsonCodec<TimeBoundary> timeBoundaryJsonCodec)
+    {
+        this.brokersForTableJsonCodec = requireNonNull(brokersForTableJsonCodec, "brokers for table json codec is null");
+        this.routingTablesJsonCodec = requireNonNull(routingTablesJsonCodec, "routing tables json codec is null");
+        this.timeBoundaryJsonCodec = requireNonNull(timeBoundaryJsonCodec, "time boundary json codec is null");
+        this.tablesJsonCodec = requireNonNull(tablesJsonCodec, "json codec is null");
+
+        this.pinotConfig = requireNonNull(pinotConfig, "pinotConfig is null");
```

Comment:
It's generally unsafe to hold on to the provided config objects as they are mutable. Extract any values needed by this class and store it in local fields.